Release/include/cpprest/ws_msg.h

```diff
@@ -60,10 +60,23 @@ class websocket_outgoing_message
 public:
 #if !defined(__cplusplus_winrt)
     /// <summary>
-    /// Sets a the outgoing message to be an unsolicited pong message.
+    /// Sets the outgoing message to be a ping message.
     /// This is useful when the client side wants to check whether the server is alive.
     /// </summary>
-    void set_pong_message() { this->set_message_pong(); }
+    /// <param name="data">UTF-8 String containing the optional ping message.</param>
+    void set_ping_message(const std::string& data = "")
```

Comment:
A ping message may or may not have a payload. But you still need a way to make the outgoing message into a ping message?

Reply:
I see. In that case it is more efficient to use `std::string()` for this than `""`.

Reply:
I’ll change tomorrow

Reply:
Just as efficient if not more so, depending on compiler, would be to use `{}` hich avoids repeating the type.

https://godbolt.org/z/ausWRL

Reply:
> Just as efficient if not more so, depending on compiler, would be to use `{}` which avoids repeating the type.
> 
> https://godbolt.org/z/ausWRL

(1) That shows the `string()` version being more efficient, since it's pushing the "" constant on the stack in the caller. But that is with the optimizer off so the results don't super matter for that, we're talking 'small cheese'.

(2) If you include the caller (which is all that matters when you're talking about default arguments) and turn on the optimizer either string() or {} (both of which call std::string's default constructor) crush "": https://godbolt.org/z/bjGgME

```
$T1 = 32
__$ArrayPad$ = 64
unsigned __int64 meow2(void) PROC                           ; meow2, COMDAT
$LN104:
        sub     rsp, 88                             ; 00000058H
        mov     rax, QWORD PTR __security_cookie
        xor     rax, rsp
        mov     QWORD PTR __$ArrayPad$[rsp], rax
        xor     r8d, r8d
        mov     QWORD PTR $T1[rsp+16], 0
        lea     rdx, OFFSET FLAT:`string'
        mov     QWORD PTR $T1[rsp+24], 15
        lea     rcx, QWORD PTR $T1[rsp]
        mov     BYTE PTR $T1[rsp], 0
        call    std::basic_string<char,std::char_traits<char>,std::allocator<char> > & std::basic_string<char,std::char_traits<char>,std::allocator<char> >::assign(char const * const,unsigned __int64) ; std::basic_string<char,std::char_traits<char>,std::allocator<char> >::assign
        mov     rdx, QWORD PTR $T1[rsp+24]
        cmp     rdx, 16
        jb      SHORT $LN74@meow2
        mov     rcx, QWORD PTR $T1[rsp]
        inc     rdx
        mov     rax, rcx
        cmp     rdx, 4096               ; 00001000H
        jb      SHORT $LN84@meow2
        mov     rcx, QWORD PTR [rcx-8]
        add     rdx, 39                             ; 00000027H
        sub     rax, rcx
        add     rax, -8
        cmp     rax, 31
        ja      SHORT $LN102@meow2
$LN84@meow2:
        call    void operator delete(void *,unsigned __int64)               ; operator delete
$LN74@meow2:
        mov     rax, QWORD PTR $T1[rsp+16]
        mov     rcx, QWORD PTR __$ArrayPad$[rsp]
        xor     rcx, rsp
        call    __security_check_cookie
        add     rsp, 88                             ; 00000058H
        ret     0
$LN102@meow2:
        call    _invalid_parameter_noinfo_noreturn
        int     3
$LN100@meow2:
unsigned __int64 meow2(void) ENDP                           ; meow2
```

for "" vs.

```
unsigned __int64 purr2(void) PROC                           ; purr2, COMDAT
        xor     eax, eax
        ret     0
unsigned __int64 purr2(void) ENDP                           ; purr2
```

(I am the std::basic_string maintainer for MSVC -- I can tell you way more about that than about any of this HTTP stuff :) )

Reply:
Sorry I first read your response as saying "`""` is just as efficient" :P

Reply:
So what do we choose finally? ^

Reply:
Either `string()` or `{}` both call the `basic_string::basic_string()` constructor and are equivalent. `""` calls the `basic_string::basic_string(const CharT*)` constructor which is less efficient (there's a strlen and a memory allocation in there; even though for "" it won't ever actually allocate it's difficult for optimizers to prove this unless the universe happens to get inlined).

Reply:
Ok then we can let {}